midiproc/main.c

```diff
@@ -0,0 +1,463 @@
+    bytes_read = Reader_BytesRead(reader);
+    DeleteReader(reader);
+    Buffer_Shift(buf, bytes_read);
+
+    // Send acknowledgement back that the command has completed.
+    if (!WriteInt16(buffer, sizeof(buffer), MIDIPIPE_PACKET_TYPE_ACK))
+    {
+        goto fail;
+    }
+
+    WriteFile(midi_process_out, buffer, sizeof(buffer),
+              &bytes_written, NULL);
+
+    return true;
+
+fail:
+    // We did not read a complete packet.  Delete our reader and try again
+    // with more data.
+    DeleteReader(reader);
+    return false;
+}
+
+//
+// The main pipe "listening" loop
+//
+boolean ListenForever()
+{
+    BOOL wok = FALSE;
+    CHAR pipe_buffer[8192];
+    DWORD pipe_buffer_read = 0;
+
+    boolean ok = false;
+    buffer_t *buffer = NewBuffer();
+
+    for (;;)
+    {
+        // Wait until we see some data on the pipe.
+        wok = PeekNamedPipe(midi_process_in, NULL, 0, NULL,
+                            &pipe_buffer_read, NULL);
+        if (!wok)
+        {
+            break;
+        }
+        else if (pipe_buffer_read == 0)
+        {
+            SDL_Delay(1);
+            continue;
+        }
+
+        // Read data off the pipe and add it to the buffer.
+        wok = ReadFile(midi_process_in, pipe_buffer, sizeof(pipe_buffer),
+                       &pipe_buffer_read, NULL);
+        if (!wok)
+        {
+            break;
+        }
+
+        ok = Buffer_Push(buffer, pipe_buffer, pipe_buffer_read);
+        if (!ok)
+        {
+            break;
+        }
+
+        do
+        {
+            // Read messages off the buffer until we can't anymore.
+            ok = ParseMessage(buffer);
+        } while (ok);
+    }
+
+    return false;
+}
+
+//=============================================================================
+//
+// Main Program
+//
+
+//
+// InitSDL
+//
+// Start up SDL and SDL_mixer.
+//
+boolean InitSDL()
+{
+    if (SDL_Init(SDL_INIT_AUDIO) == -1)
+    {
+        return false;
+    }
+
+    if (Mix_OpenAudioDevice(snd_samplerate, MIX_DEFAULT_FORMAT, 2, 2048, NULL, SDL_AUDIO_ALLOW_FREQUENCY_CHANGE) < 0)
+    {
+        return false;
+    }
+
+    atexit(ShutdownSDL);
+
+    return true;
+}
+
+//
+// InitPipes
+//
+// Ensure that we can communicate.
+//
+void InitPipes(HANDLE in, HANDLE out)
+{
+    midi_process_in = in;
+    midi_process_out = out;
+
+    atexit(FreePipes);
+}
+
+//
+// main
+//
+// Application entry point.
+//
+int main(int argc, char *argv[])
+{
+    HANDLE in, out;
+
+    // Make sure we're not launching this process by itself.
+    if (argc < 5)
+    {
+        MessageBox(NULL, TEXT("This program is tasked with playing Native ")
+                   TEXT("MIDI music, and is intended to be launched by ")
+                   TEXT(PROJECT_NAME) TEXT("."),
+                   TEXT(PROJECT_STRING), MB_OK | MB_ICONASTERISK);
+
+        return EXIT_FAILURE;
+    }
+
+    // Make sure our Choccolate Doom and midiproc version are lined up.
```

Comment:
Why have you commented this out?

Reply:
Fix in https://github.com/fabiangreffrath/woof/pull/100/commits/33e2d0c750b38b95af6d37f420aa86a53363b7e4